File: src/python/xerxes/runtime/nudges.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Periodic nudges that shape agent self-improvement.

Two built-in rules ship by default:

    * :class:`MemoryNudge` — every ``interval`` turns, if no memory writes
      have happened since the last fire and the conversation contains
      durable info, inject a one-shot system note suggesting
      ``save_memory(...)``.
    * :class:`SkillNudge` — after a turn with ``>= threshold`` successful
      tool calls, suggest ``skill_manage`` to capture the pattern.

Nudges are *suggestions*, not interruptions: each fires by appending a
short system-role message that the LLM reads on the *next* turn. Rules
can be silenced individually via :meth:`NudgeManager.disable`.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Any


@dataclass
class NudgeContext:
    """Snapshot of session state the nudge rules consult.

    Attributes:
        turn_index: 0-based count of completed turns in the session.
        tool_calls_this_turn: Tool calls issued during the just-finished turn.
        successful_tool_calls_this_turn: Subset of ``tool_calls_this_turn``
            that returned without error.
        memory_writes_this_turn: Memory-writing calls (``save_memory`` etc.)
            during this turn.
        memory_writes_since_last_fire: Manager-tracked rolling count used
            for rate-limiting :class:`MemoryNudge`.
        last_user_message: Most recent user turn text, or ``""``.
        last_assistant_message: Most recent assistant turn text, or ``""``.
        metadata: Free-form bag for rule-specific extras.
    """

    turn_index: int
    tool_calls_this_turn: int = 0
    successful_tool_calls_this_turn: int = 0
    memory_writes_this_turn: int = 0
    memory_writes_since_last_fire: int = 0
    last_user_message: str = ""
    last_assistant_message: str = ""
    metadata: dict[str, Any] = field(default_factory=dict)


class NudgeRule(ABC):
    """Abstract base class for nudge rules.

    Subclasses set a ``name`` class attribute and implement
    :meth:`should_fire` and :meth:`message`.
    """

    name: str = ""

    @abstractmethod
    def should_fire(self, ctx: NudgeContext) -> bool:
        """Return ``True`` when this rule should emit a nudge for ``ctx``."""
        ...

    @abstractmethod
    def message(self, ctx: NudgeContext) -> str:
        """Return the system-role message to inject before the next turn."""
        ...


class MemoryNudge(NudgeRule):
    """Reminds the agent to persist durable info via ``save_memory``."""

    name = "memory"

    DURABLE_HINTS = ("decided", "prefers", "will always", "should never", "username", "deadline", "remember")
    """Substring markers that suggest the conversation produced durable info."""

    def __init__(self, *, interval: int = 8) -> None:
        """Fire at most once every ``interval`` turns (default ``8``)."""
        self.interval = int(interval)

    def should_fire(self, ctx: NudgeContext) -> bool:
        """Return ``True`` when the interval is up and durable hints are present."""
        if ctx.memory_writes_since_last_fire > 0:
            return False
        if (ctx.turn_index + 1) % self.interval != 0:
            return False
        blob = f"{ctx.last_user_message} {ctx.last_assistant_message}".lower()
        return any(hint in blob for hint in self.DURABLE_HINTS)

    def message(self, ctx: NudgeContext) -> str:
        """Return the canned save-to-memory suggestion."""
        return (
            "[NUDGE] You haven't written to memory in a while and the recent turn looks like it had "
            "durable user info (preferences, decisions, deadlines). Consider calling "
            "`save_memory(content=...)` so this doesn't get lost across sessions."
        )


class SkillNudge(NudgeRule):
    """Suggests formalising a heavy tool chain into a reusable skill."""

    name = "skill"

    def __init__(self, *, threshold: int = 6) -> None:
        """Fire when a single turn made at least ``threshold`` successful tool calls."""
        self.threshold = int(threshold)

    def should_fire(self, ctx: NudgeContext) -> bool:
        """Return ``True`` when the successful tool count crosses ``threshold``."""
        return ctx.successful_tool_calls_this_turn >= self.threshold

    def message(self, ctx: NudgeContext) -> str:
        """Return the canned skill-capture suggestion."""
        return (
            "[NUDGE] The just-finished turn used many tools successfully. If this pattern is likely "
            "to recur, consider `skill_manage(intent='create', name=..., body=...)` to capture it "
            "as a reusable skill."
        )


class NudgeManager:
    """Coordinate a list of :class:`NudgeRule` instances and their fired state.

    Each call to :meth:`check` returns the messages every active rule wants to
    inject before the next turn. Per-rule fire counts are tracked so external
    callers can rate-limit display or reset state.
    """

    def __init__(self, rules: list[NudgeRule] | None = None) -> None:
        """Initialise with ``rules`` or the default ``[MemoryNudge, SkillNudge]``."""
        self._rules = rules or [MemoryNudge(), SkillNudge()]
        self._disabled: set[str] = set()
        self._fired_count: dict[str, int] = {}

    @property
    def rules(self) -> list[NudgeRule]:
        """Return a shallow copy of the registered rule list."""
        return list(self._rules)

    def disable(self, rule_name: str) -> None:
        """Silence the rule with ``rule_name`` until :meth:`enable` is called."""
        self._disabled.add(rule_name)

    def enable(self, rule_name: str) -> None:
        """Re-enable a previously disabled rule."""
        self._disabled.discard(rule_name)

    def disabled(self) -> set[str]:
        """Return a copy of the currently disabled rule names."""
        return set(self._disabled)

    def fired_count(self, rule_name: str) -> int:
        """Return how many times the named rule has fired so far."""
        return self._fired_count.get(rule_name, 0)

    def check(self, ctx: NudgeContext) -> list[tuple[str, str]]:
        """Return ``[(rule_name, message), ...]`` for every rule that fires."""
        out: list[tuple[str, str]] = []
        for rule in self._rules:
            if rule.name in self._disabled:
                continue
            if rule.should_fire(ctx):
                out.append((rule.name, rule.message(ctx)))
                self._fired_count[rule.name] = self._fired_count.get(rule.name, 0) + 1
        return out


__all__ = ["MemoryNudge", "NudgeContext", "NudgeManager", "NudgeRule", "SkillNudge"]

File: src/typescript/src/runtime/nudges.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

/** Input values for a completed turn that nudge rules inspect. */
export interface NudgeContextOptions {
  readonly lastAssistantMessage?: string
  readonly lastUserMessage?: string
  readonly memoryWritesSinceLastFire?: number
  readonly memoryWritesThisTurn?: number
  readonly metadata?: Readonly<Record<string, unknown>>
  readonly successfulToolCallsThisTurn?: number
  readonly toolCallsThisTurn?: number
  readonly turnIndex: number
}

/** Immutable snapshot of the session state available to nudge rules. */
export class NudgeContext {
  readonly lastAssistantMessage: string
  readonly lastUserMessage: string
  readonly memoryWritesSinceLastFire: number
  readonly memoryWritesThisTurn: number
  readonly metadata: Readonly<Record<string, unknown>>
  readonly successfulToolCallsThisTurn: number
  readonly toolCallsThisTurn: number
  readonly turnIndex: number

  constructor(options: NudgeContextOptions) {
    this.turnIndex = nonNegativeInteger(options.turnIndex, 'turnIndex')
    this.toolCallsThisTurn = nonNegativeInteger(options.toolCallsThisTurn ?? 0, 'toolCallsThisTurn')
    this.successfulToolCallsThisTurn = nonNegativeInteger(
      options.successfulToolCallsThisTurn ?? 0,
      'successfulToolCallsThisTurn',
    )
    this.memoryWritesThisTurn = nonNegativeInteger(options.memoryWritesThisTurn ?? 0, 'memoryWritesThisTurn')
    this.memoryWritesSinceLastFire = nonNegativeInteger(
      options.memoryWritesSinceLastFire ?? 0,
      'memoryWritesSinceLastFire',
    )
    this.lastUserMessage = options.lastUserMessage ?? ''
    this.lastAssistantMessage = options.lastAssistantMessage ?? ''
    this.metadata = Object.freeze({ ...(options.metadata ?? {}) })
    Object.freeze(this)
  }
}

/** A named suggestion rule evaluated after a completed agent turn. */
export interface NudgeRule {
  readonly name: string
  message(context: NudgeContext): string
  shouldFire(context: NudgeContext): boolean
}

export type NudgeResult = readonly [name: string, message: string]

/** Reminds the agent to persist durable information via save_memory. */
export class MemoryNudge implements NudgeRule {
  static readonly DURABLE_HINTS = Object.freeze([
    'decided',
    'prefers',
    'will always',
    'should never',
    'username',
    'deadline',
    'remember',
  ])

  readonly name = 'memory'
  readonly interval: number

  constructor(options: { readonly interval?: number } = {}) {
    this.interval = positiveInteger(options.interval ?? 8, 'interval')
  }

  shouldFire(context: NudgeContext): boolean {
    if (context.memoryWritesSinceLastFire > 0) return false
    if ((context.turnIndex + 1) % this.interval !== 0) return false
    const conversation = `${context.lastUserMessage} ${context.lastAssistantMessage}`.toLowerCase()
    return MemoryNudge.DURABLE_HINTS.some(hint => conversation.includes(hint))
  }

  message(_context: NudgeContext): string {
    return '[NUDGE] You haven\'t written to memory in a while and the recent turn looks like it had '
      + 'durable user info (preferences, decisions, deadlines). Consider calling '
      + '`save_memory(content=...)` so this doesn\'t get lost across sessions.'
  }
}

/** Suggests capturing a successful, tool-heavy turn as a reusable skill. */
export class SkillNudge implements NudgeRule {
  readonly name = 'skill'
  readonly threshold: number

  constructor(options: { readonly threshold?: number } = {}) {
    this.threshold = nonNegativeInteger(options.threshold ?? 6, 'threshold')
  }

  shouldFire(context: NudgeContext): boolean {
    return context.successfulToolCallsThisTurn >= this.threshold
  }

  message(_context: NudgeContext): string {
    return '[NUDGE] The just-finished turn used many tools successfully. If this pattern is likely '
      + 'to recur, consider `skill_manage(intent=\'create\', name=..., body=...)` to capture it '
      + 'as a reusable skill.'
  }
}

/** Coordinates rule enablement and per-rule fire counts for one session/runtime. */
export class NudgeManager {
  private readonly disabledRules = new Set<string>()
  private readonly firedCounts = new Map<string, number>()
  private readonly ruleList: NudgeRule[]

  constructor(rules: readonly NudgeRule[] = []) {
    this.ruleList = rules.length ? [...rules] : [new MemoryNudge(), new SkillNudge()]
  }

  /** Return a shallow copy so callers cannot reorder the manager's rule list. */
  get rules(): NudgeRule[] {
    return [...this.ruleList]
  }

  disable(ruleName: string): void {
    this.disabledRules.add(ruleName)
  }

  enable(ruleName: string): void {
    this.disabledRules.delete(ruleName)
  }

  /** Return a copy of disabled names rather than exposing mutable manager state. */
  disabled(): Set<string> {
    return new Set(this.disabledRules)
  }

  firedCount(ruleName: string): number {
    return this.firedCounts.get(ruleName) ?? 0
  }

  /** Return every active suggestion that should be injected before the next turn. */
  check(context: NudgeContext): NudgeResult[] {
    const results: NudgeResult[] = []
    for (const rule of this.ruleList) {
      if (this.disabledRules.has(rule.name) || !rule.shouldFire(context)) continue
      results.push([rule.name, rule.message(context)])
      this.firedCounts.set(rule.name, this.firedCount(rule.name) + 1)
    }
    return results
  }
}

function nonNegativeInteger(value: number, name: string): number {
  if (!Number.isSafeInteger(value) || value < 0) {
    throw new RangeError(name + ' must be a non-negative safe integer')
  }
  return value
}

function positiveInteger(value: number, name: string): number {
  if (!Number.isSafeInteger(value) || value <= 0) {
    throw new RangeError(name + ' must be a positive safe integer')
  }
  return value
}
